modules/features2d/src/matchers.cpp

```diff
@@ -613,7 +613,7 @@ void DescriptorMatcher::checkMasks( InputArrayOfArrays _masks, int queryDescript
             {
                 int rows = trainDescCollection[i].empty() ? utrainDescCollection[i].rows : trainDescCollection[i].rows;
                     CV_Assert( masks[i].rows == queryDescriptorsCount &&
-                        (masks[i].cols == rows || masks[i].cols == rows) &&
+                        (masks[i].rows == rows || masks[i].cols == rows) &&
```

Comment:
masks[i].rows is already checked at line 615 so only masks[i].cols should be checked here. masks[i] should be a matrix with query descriptors count rows and train descriptors count columns(in accordance with [parameters description](http://docs.opencv.org/3.1.0/db/d39/classcv_1_1DescriptorMatcher.html#a8737756123c4aba88ae5fc15982e0dee) for DescriptorMatcher::Match)